### PR TITLE
Add 'auto_add_group_owner_membership' setting

### PR DESCRIPTION
**Pull request by bboro**
Case study:

Sometimes website may be creating entities programmatically, or just plain do not desire automatic membership creation for the group creator. In such cases, it's quite difficult to prevent this from happening due to the creation being in `og_entity_insert`.

Adding new global setting: 'Automatically add creators to the group' 

image

* When the setting is ON (default) membership is created
  automatically for group owners/creators.
* When the setting is OFF membership is not created automatically
  for group owners/creators.

Changelist:

* Added config schema changes
* Added default installation config changes
* Added update hook to update existing configurations
* Added the setting in the AdminSettingsForm
* Updated og_entity_insert to be aware of the new config setting
* Updated tests

### CODE REVIEW

**Comment by bboro** [comment]
This requirement was originally stemming from https://github.com/Gizra/jep-portal/issues/663 where we don't want to create memberships automatically for group owners.

File: og_ui/src/Form/AdminSettingsForm.php [hunk 1]
```diff
@@ -133,6 +132,13 @@ public function buildForm(array $form, FormStateInterface $form_state) {
       }
     }
 
+    $form['auto_add_owners'] = [
```
Comment (amitaibu) [inline og_ui/src/Form/AdminSettingsForm.php]: How about a more explicit name, like

```suggestion
    $form['auto_add_membership_to_group_owner'] = [
```

Comment (bboro) [reply og_ui/src/Form/AdminSettingsForm.php]: Renamed to `auto_add_group_owner_membership`

**Comment by bboro** [comment]
Looks like some PHPCS failures, but looks like fixes are coming in another PR. I had a quick look at my test failures and seems like it's nothing introduced from my side, but I may have easily missed it.

**Comment by pfrenssen** [comment]
I did a code review and everything looks good to me. Thanks for adding test coverage!

I did not test it out in the UI though, if someone can confirm this works as expected when setting the option in the UI this is good to go.

**Comment by amitaibu** [comment]
I've confirmed manually the feature works

## Enabled

image


## Disabled

image


I think as a follow up we should show "No members in the group yet" message, as we've never had up until now empty groups. @bboro care to tackle this?

**Review by pfrenssen — APPROVED** [review]
Thanks for the test and the screenshots!

Approving, but we will have to wait for #700 to be in before merging this. We're affected by a test failure which is being fixed in that PR.

**Comment by bboro** [comment]
@pfrenssen thanks for the review!

@amitaibu sure thing I'll make some time to work on this tomorrow. Question about the implementation: Since the members listing is a view, how should we handle the view configuration on existing installations? Do we just leave them be since they may have a custom empty message set already?

**Comment by amitaibu** [comment]
> how should we handle the view configuration on existing installations?

You can change the existing one. As it was not (easily) possibly to have no-members, I don't think it's a problem.

File: og.install [hunk 2]
```diff
@@ -93,3 +94,48 @@ function og_update_8002() {
...
+    'table' => 'views',
+    'field' => 'area',
+    'relationship' => 'none',
+    'group_type' => 'group',
+    'admin_label' => 'No results text',
+    'empty' => TRUE,
+    'tokenize' => FALSE,
+    'content' => [
+      'value' => 'No members in the group yet.',
+      'format' => 'basic_html',
```
Comment (bboro) [inline og.install]: Is it ok to assume basic_html is an existing format here? (Also same in the optional installation view config) 

If not, does anyone have any pointers on how this is usually handled? Perhaps I should use a different plugin/field?

Comment (amitaibu) [reply og.install]: I don't think we need an upgrade path for Views config.

Comment (bboro) [reply og.install]: > I don't think we need an upgrade path for Views config.

@amitaibu so we don't want to update existing installations and won't need this update hook? I'm not sure I understand exactly what you mean.

Comment (amitaibu) [reply og.install]: Correct, no upgrade hook for Views

Comment (bboro) [reply og.install]: Thanks! I'll remove

Comment (bboro) [reply og.install]: FYI I'm also setting the empty text to unfiltered text in the installationg config yml instead of trying to force text formats

**Comment by MPParsley** [comment]
#700 is fixed, updating branch.

File: og.install [hunk 3]
```diff
@@ -9,6 +9,7 @@ declare(strict_types = 1);
 
 use Drupal\Core\Field\BaseFieldDefinition;
 use Drupal\Core\StringTranslation\TranslatableMarkup;
+use Drupal\views\Views;
```
Comment (MPParsley) [inline og.install]: Unused use statement:

```suggestion
```

File: tests/src/Kernel/GroupManagerSubscriptionTest.php [hunk 4]
```diff
@@ -117,6 +117,26 @@ public function testGroupManagerSubscription($group_has_owner, $membership_is_ov
...
+    $config->set('auto_add_group_owner_membership', $auto_add_group_owner_membership);
+    $config->save();
+    $group = Node::create([
+      'title' => 'membership is not overridden',
+      'type' => 'group',
+      'uid' => $this->owner->id(),
+    ]);
+    $group->save();
+    $membership = $this->membershipManager->getMembership($group, $this->owner->id());
+    // Membership should be created only when auto_add_group_owner_membership is enabled.
```
Comment (MPParsley) [inline tests/src/Kernel/GroupManagerSubscriptionTest.php]: Line exceeds 80 characters; contains 89 characters 

```suggestion
    // Membership should be created only when auto_add_group_owner_membership is
    // enabled.
```

**Comment by MPParsley** [comment]
Fixed CS issues:

```
FILE: /home/travis/build/Gizra/og/og.install
----------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
----------------------------------------------------------------------
 12 | WARNING | [x] Unused use statement
----------------------------------------------------------------------
PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------
FILE: ...me/travis/build/Gizra/og/tests/src/Kernel/GroupManagerSubscriptionTest.php
--------------------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
--------------------------------------------------------------------------------
 136 | WARNING | Line exceeds 80 characters; contains 89 characters
--------------------------------------------------------------------------------
```

File: tests/src/Kernel/GroupManagerSubscriptionTest.php [hunk 5]
```diff
@@ -176,4 +197,24 @@ public static function groupManagerSubscriptionProvider() {
...
+   * Data provider for ::testAutoAddOwnersSetting().
+   *
+   * @return array[]
+   *   Array of data with values for the 'auto_add_group_owner_membership'
+   *   config key.
+   */
+  public static function autoAddOwnersSettingProvider() {
+    return [
+      [
+        // Auto Add owners.
```
Comment (MPParsley) [inline tests/src/Kernel/GroupManagerSubscriptionTest.php]:
```suggestion
        // Auto add owners.
```

File: config/optional/views.view.og_members_overview.yml [hunk 6]
```diff
@@ -430,7 +428,18 @@ display:
...
+        area_text_custom:
+          id: area_text_custom
+          table: views
+          field: area_text_custom
+          relationship: none
+          group_type: group
+          admin_label: 'No results text'
+          empty: true
+          tokenize: false
+          content: 'No members in the group yet.'
```
Comment (MPParsley) [inline config/optional/views.view.og_members_overview.yml]:
```suggestion
          content: 'There are no members in this group.'
```

```suggestion
          content: 'There are no members that belong to this group.'
```

**Review by MPParsley — COMMENTED** [review]
Test are green, just added a small grammar suggestion.

Should we make the default role configurable?

**Comment by bboro** [comment]
Rerolled against latest 8.x-1.x

**Review by MPParsley — APPROVED** [review]

**Pull request merged**